vnext/Shared/Networking/WinRTWebSocketResource.cpp

```diff
@@ -79,10 +82,356 @@ auto resume_in_queue(const Mso::DispatchQueue &queue) noexcept {
   return awaitable{queue};
 } // resume_in_queue
 
+DispatchQueue GetCurrentOrSerialQueue() noexcept {
+  auto queue = DispatchQueue::CurrentQueue();
+  if (!queue)
+    queue = DispatchQueue::MakeSerialQueue();
+
+  return queue;
+}
+
 } // namespace
 
 namespace Microsoft::React::Networking {
 
+#pragma region WinRTWebSocketResource2
+
+WinRTWebSocketResource2::WinRTWebSocketResource2(
+    IMessageWebSocket &&socket,
+    IDataWriter &&writer,
+    vector<ChainValidationResult> &&certExceptions,
+    DispatchQueue callingQueue)
+    : m_socket{std::move(socket)},
+      m_writer(std::move(writer)),
+      m_readyState{ReadyState::Connecting},
+      m_connectPerformed{CreateEvent(/*attributes*/ nullptr, /*manual reset*/ true, /*state*/ false, /*name*/ nullptr)},
+      m_callingQueue{callingQueue} {
+  for (const auto &certException : certExceptions) {
+    m_socket.Control().IgnorableServerCertificateErrors().Append(certException);
+  }
+}
+
+// private
+WinRTWebSocketResource2::WinRTWebSocketResource2(
+    IMessageWebSocket &&socket,
+    vector<ChainValidationResult> &&certExceptions)
+    : WinRTWebSocketResource2(
+          std::move(socket),
+          DataWriter{socket.OutputStream()},
+          std::move(certExceptions),
+          GetCurrentOrSerialQueue()) {}
+
+WinRTWebSocketResource2::WinRTWebSocketResource2(vector<ChainValidationResult> &&certExceptions)
+    : WinRTWebSocketResource2(MessageWebSocket{}, std::move(certExceptions)) {}
+
+WinRTWebSocketResource2::~WinRTWebSocketResource2() noexcept /*override*/
+{}
+
+void WinRTWebSocketResource2::Fail(string &&message, ErrorType type) noexcept {
+  auto self = shared_from_this();
+
+  self->m_readyState = ReadyState::Closed;
```

Comment:
>m_readyState [](http://example.com/codeflow?start=8&length=12)

It is important to ensure that this field is only accessed only from the dispatch queue for the synchronization. It does not seem that the Fail function is accessed only from the DQ.

Reply:
This has been fixed in the latest iteration:
```
self->m_backgroundQueue.Post([self]() { self->m_readyState = ReadyState::Closed; });
```